cli/cli-client/src/main/java/org/infinispan/cli/user/UserTool.java

```diff
@@ -0,0 +1,370 @@
+         this.serverRoot = serverRoot;
+      } else {
+         String serverHome = System.getProperty("infinispan.server.home.path");
+         Path serverHomePath = serverHome == null ? Paths.get("") : Paths.get(serverHome);
+         if (serverRoot == null) {
+            this.serverRoot = serverHomePath.resolve("server");
+         } else {
+            this.serverRoot = serverHomePath.resolve(serverRoot);
+         }
+      }
+
+      if (usersFile == null) {
+         this.usersFile = this.serverRoot.resolve("conf").resolve(DEFAULT_USERS_PROPERTIES_FILE);
+      } else if (usersFile.isAbsolute()) {
+         this.usersFile = usersFile;
+      } else {
+         this.usersFile = this.serverRoot.resolve("conf").resolve(usersFile);
+      }
+      if (groupsFile == null) {
+         this.groupsFile = this.serverRoot.resolve("conf").resolve(DEFAULT_GROUPS_PROPERTIES_FILE);
+      } else if (groupsFile.isAbsolute()) {
+         this.groupsFile = groupsFile;
+      } else {
+         this.groupsFile = this.serverRoot.resolve("conf").resolve(groupsFile);
+      }
+      load();
+   }
+
+   private void installSecurityProvider() {
+      WildFlyElytronPasswordProvider instance = WildFlyElytronPasswordProvider.getInstance();
+      if (java.security.Security.getProvider(instance.getName()) == null) {
+         java.security.Security.insertProviderAt(instance, 1);
+      }
+   }
+
+
+   public void reload() {
+      this.realm = null;
+      this.encryption = Encryption.DEFAULT;
+      load();
+   }
+
+   private void load() {
+      if (Files.exists(usersFile)) {
+         try (BufferedReader reader = Files.newBufferedReader(usersFile, StandardCharsets.UTF_8)) {
+            String currentLine;
+            while ((currentLine = reader.readLine()) != null) {
+               final String trimmed = currentLine.trim();
+               if (trimmed.startsWith(COMMENT_PREFIX1) && trimmed.contains(REALM_COMMENT_PREFIX)) {
+                  // this is the line that contains the realm name.
+                  int start = trimmed.indexOf(REALM_COMMENT_PREFIX) + REALM_COMMENT_PREFIX.length();
+                  int end = trimmed.indexOf(COMMENT_SUFFIX, start);
+                  if (end > -1) {
+                     realm = trimmed.substring(start, end);
+                  }
+               } else if (trimmed.startsWith(COMMENT_PREFIX1) && trimmed.contains(ALGORITHM_COMMENT_PREFIX)) {
+                  // this is the line that contains the algorithm name.
+                  int start = trimmed.indexOf(ALGORITHM_COMMENT_PREFIX) + ALGORITHM_COMMENT_PREFIX.length();
+                  int end = trimmed.indexOf(COMMENT_SUFFIX, start);
+                  if (end > -1) {
+                     encryption = Encryption.valueOf(trimmed.substring(start, end).toUpperCase());
+                  }
+               } else {
+                  if (!(trimmed.startsWith(COMMENT_PREFIX1) || trimmed.startsWith(COMMENT_PREFIX2))) {
```

Comment:
well, you could have paid the cost of reading the file twice instead of writing / copying the parser.

Reply:
You mean one pass to find the comments and then use the standard properties loader ? True, but that code is what the actual realm loader uses, so it's tested and I feel safer :)